### PR TITLE
chore: switch from pipenv to uv

### PR DESCRIPTION
**Pull request by leseb**
uv from the people who wrote ruff is an extremely fast Python package and project manager.

## Description

5989502 chore: switch from pipenv to uv

commit 598950220c99564a5a4f32463b45d1a9b9c23f30
Author: Sébastien Han <seb@redhat.com>
Date:   Mon Dec 9 15:33:46 2024 +0100

    chore: switch from pipenv to uv
    
    uv from the people who wrote ruff is an extremely fast Python package
    and project manager.
    
    Signed-off-by: Sébastien Han <seb@redhat.com>

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Review by tumido — COMMENTED** [review]
Love it. ❤️  However we should announce that this PR would require everybody to migrate from `pipenv` to `uv`. It also changes how you work with the repo depending on if you checkout to `phase-1`/`phase-2` or `main`...

Maybe we should keep `Pipfile`/`Pipfile.lock` around for compatibility sake and advice to use `uv` in the future. And we should also consider backporting to `phase-1`/`phase-2` before removing Pipenv from the picture?

**Comment by leseb** [comment]
> Love it. ❤️ However we should announce that this PR would require everybody to migrate from `pipenv` to `uv`. It also changes how you work with the repo depending on if you checkout to `phase-1`/`phase-2` or `main`...

Correct, I'm planning backporting this once we agree to merge it.

> 
> Maybe we should keep `Pipfile`/`Pipfile.lock` around for compatibility sake and advice to use `uv` in the future. And we should also consider backporting to `phase-1`/`phase-2` before removing Pipenv from the picture?

While keeping Pipfile/Pipfile.lock around for compatibility seems fine, maintaining two dependency management tools would introduce overhead. Given the relatively small contributor base, I lean towards a clean break from Pipenv. Breaking early will encourage contributors to adopt uv sooner and reduce the risk of confusion or delays. I can write a small contributing guide that reflect how to build a venv with uv.

What do you think?

**Comment by tumido** [comment]
I'm fine with that... I can manage the slight inconvenience changing the venv provider when rebasing my PRs (in case my branch is on `Pipenv` and main is on `uv` with different venvs...). Not sure how much inconvenient it will be for others. 🙂 

Clean break is fine with me, I would love to hear what others think. Maybe having a simple quick migration guide would be helpful. I see multiple options for that as well:
- a markdown doc in repo / README entry (which we would need to clean up after some time, once it's no longer relevant)
- as a documentation in this PR linked from README or something
- a separate GH issue that we can pin of the time being and retire once we are assured everybody has migrated...

Either option is fine. Whatever is easier to manage. 🙂

**Comment by leseb** [comment]
> I'm fine with that... I can manage the slight inconvenience changing the venv provider when rebasing my PRs (in case my branch is on `Pipenv` and main is on `uv` with different venvs...). Not sure how much inconvenient it will be for others. 🙂
> 
> Clean break is fine with me, I would love to hear what others think. Maybe having a simple quick migration guide would be helpful. I see multiple options for that as well:
> 
> * a markdown doc in repo / README entry (which we would need to clean up after some time, once it's no longer relevant)
> * as a documentation in this PR linked from README or something
> * a separate GH issue that we can pin of the time being and retire once we are assured everybody has migrated...
> 
> Either option is fine. Whatever is easier to manage. 🙂

In terms of migration, there is not much to do despite installing `uv` and getting familiar with the CLI by going through their [docs](https://docs.astral.sh/uv/). Do we really need to write something for this?

**Comment by tumido** [comment]
I guess it can be as simple as adding something like this to the README? WDYT?


````md
## Developer setup

To collaborate on this repository, please follow these steps:
1. Install [uv](https://docs.astral.sh/uv/getting-started/installation/)
2. Run following commands to prepare your local environment
   ```
   uv venv && uv sync
   ```
````

**Comment by tumido** [comment]
In addition we need to communicate that the day to day workflow changes from `pipenv shell` to `source .venv/bin/activate`. 🤷 🙂

File: README.md [hunk 1]
```diff
@@ -189,3 +189,16 @@ To generate the `standalone.py` script, run the following command ([ruff](https:
...
+## Developer setup
+
+To collaborate on this repository, please follow these steps:
+
+1. Install [uv](https://docs.astral.sh/uv/getting-started/installation/)
+2. Run following commands to prepare your local environment
+
+```bash
+uv sync
+uv venv
```
Comment (tumido) [inline README.md:202]: I've found out this command is actually not necessary. `uv sync` by itself creates a `.venv` if it's not already present. That's neat. 🙂

File: README.md [hunk 2]
```diff
@@ -189,3 +189,16 @@ To generate the `standalone.py` script, run the following command ([ruff](https:
...
 ```
+
+## Developer setup
+
+To collaborate on this repository, please follow these steps:
+
+1. Install [uv](https://docs.astral.sh/uv/getting-started/installation/)
+2. Run following commands to prepare your local environment
+
+```bash
```
Comment (tumido) [inline README.md:200]: Can we nest this code block into the 2) ?

**Review by tumido — COMMENTED** [review]
Just a few README suggestions. This should be the last time, I promise! 😄 🤷

**Review by tumido — APPROVED** [review]
Like it. 👍 Let's announce it before merging. Whenever you feel like it. 🙂 

/lgtm
/approve

**Pull request merged**